In [46]:
iris <- read.csv('iris.csv')

In [47]:
set.seed(12)
N = nrow(iris)
tr.idx = sample(1:N, size = N*2/3, replace = FALSE)

## 선형판별분석 (LDA)
### attributes in training and test

In [48]:
# attributes in training and test
iris.train <- iris[tr.idx, -5]
iris.test <- iris[-tr.idx, -5]
# target value in training and test
trainLabels <- iris[tr.idx, 5]
testLabels <- iris[-tr.idx, 5]

In [49]:
train <- iris[tr.idx, ]
test <- iris[-tr.idx, ]

In [50]:
# MASS package for LDA
library(MASS)

In [51]:
# Linear Discriminant Analysis (LDA) with training data n=100
iris.lda <- lda(Species ~ ., data = train, prior = c(1/3, 1/3, 1/3))
iris.lda

Call:
lda(Species ~ ., data = train, prior = c(1/3, 1/3, 1/3))

Prior probabilities of groups:
    setosa versicolor  virginica 
 0.3333333  0.3333333  0.3333333 

Group means:
           Sepal.Length Sepal.Width Petal.Length Petal.Width
setosa         5.019444    3.425000     1.450000   0.2416667
versicolor     5.963636    2.781818     4.278788   1.3242424
virginica      6.532258    2.974194     5.541935   2.0290323

Coefficients of linear discriminants:
                    LD1        LD2
Sepal.Length -0.6592007 -0.5763915
Sepal.Width  -1.6123822  2.3168756
Petal.Length  2.4027730 -0.5964554
Petal.Width   2.7293443  2.5978979

Proportion of trace:
  LD1   LD2 
0.993 0.007 

### Group means는 각 독립변수의 평균을 보여준다.
### LD1 = -0.66 Sepal.Length -1.61 Sepal.Width 2.4 Petal.Length 2.73 Petal.Width
### LD2 = -0.58 Sepal.Length 2.32 Sepal.Width -0.6 Petal.Length 2.6 Petal.Width
### LD1이 between-group variance의 99%를 설명
### LD2가 between-group variance의 1%를 설명

In [52]:
testpred <- predict(iris.lda, test)
testpred

$class
 [1] setosa     setosa     setosa     setosa     setosa     setosa    
 [7] setosa     setosa     setosa     setosa     setosa     setosa    
[13] setosa     setosa     versicolor versicolor versicolor versicolor
[19] versicolor virginica  versicolor versicolor versicolor versicolor
[25] versicolor versicolor versicolor versicolor versicolor versicolor
[31] versicolor virginica  virginica  virginica  virginica  virginica 
[37] virginica  virginica  virginica  virginica  virginica  virginica 
[43] virginica  virginica  virginica  virginica  virginica  virginica 
[49] virginica  virginica 
Levels: setosa versicolor virginica

$posterior
          setosa   versicolor    virginica
1   1.000000e+00 3.608279e-25 1.420546e-48
3   1.000000e+00 9.040178e-23 1.915789e-45
6   1.000000e+00 9.933481e-24 8.860313e-46
8   1.000000e+00 5.168749e-23 1.134427e-45
19  1.000000e+00 7.818955e-25 8.398049e-48
23  1.000000e+00 5.723920e-29 1.884922e-53
27  1.000000e+00 1.006326e-19 1.101634e-40
31  1.000000e+00 5.552896e-19 2.199671e-40
35  1.000000e+00 2.464986e-20 2.644751e-42
43  1.000000e+00 5.768218e-22 3.499776e-44
45  1.000000e+00 5.945388e-20 1.549913e-40
46  1.000000e+00 3.030640e-19 1.174396e-40
47  1.000000e+00 2.070108e-25 1.600401e-48
50  1.000000e+00 2.706131e-23 3.449179e-46
51  1.302091e-22 9.999572e-01 4.275865e-05
64  1.942796e-27 9.965570e-01 3.442959e-03
66  5.749910e-21 9.999895e-01 1.051618e-05
69  9.746197e-32 9.725640e-01 2.743599e-02
70  5.220382e-20 9.999994e-01 5.907064e-07
71  1.590178e-31 3.406014e-01 6.593986e-01
73  1.737872e-33 7.912964e-01 2.087036e-01
77  1.903881e-27 9.988360e-01 1.163997e-03
81  5.375894e-20 9.999995e-01 4.871061e-07
83  5.735497e-19 9.999994e-01 5.627844e-07
85  1.880286e-27 9.830395e-01 1.696050e-02
89  2.295246e-20 9.999902e-01 9.759514e-06
94  1.288790e-15 1.000000e+00 8.157443e-09
95  7.349165e-24 9.999126e-01 8.735541e-05
97  5.512047e-22 9.999730e-01 2.698388e-05
99  1.367770e-11 1.000000e+00 6.278048e-10
100 1.052821e-21 9.999843e-01 1.567059e-05
101 7.022870e-59 1.074143e-09 1.000000e+00
103 5.483972e-50 4.933870e-06 9.999951e-01
108 6.792283e-51 1.211085e-05 9.999879e-01
111 8.207017e-37 1.240063e-02 9.875994e-01
119 7.098235e-70 2.871704e-11 1.000000e+00
120 1.420440e-38 1.288745e-01 8.711255e-01
125 1.006213e-45 2.824626e-05 9.999718e-01
127 4.035247e-34 2.214762e-01 7.785238e-01
128 4.886885e-34 1.494286e-01 8.505714e-01
129 2.131457e-50 3.268111e-06 9.999967e-01
131 3.350587e-50 1.757950e-05 9.999824e-01
132 1.618984e-44 8.120881e-05 9.999188e-01
136 5.967778e-54 3.118501e-07 9.999997e-01
139 5.212379e-33 2.433934e-01 7.566066e-01
141 2.021326e-51 3.834032e-07 9.999996e-01
143 6.420658e-43 6.015733e-04 9.993984e-01
146 1.160988e-44 5.082665e-05 9.999492e-01
148 2.848267e-40 2.113569e-03 9.978864e-01
150 8.200002e-38 1.272406e-02 9.872759e-01

$x
           LD1         LD2
1   -8.6099765  0.25655128
3   -8.1028589 -0.14830927
6   -8.1859889  1.35102703
8   -8.1425409  0.02285732
19  -8.4954453  0.68663223
23  -9.4027236  1.01501676
27  -7.3563948  0.48279135
31  -7.2867088 -0.61657262
35  -7.5929062 -0.61456622
43  -7.9050986  0.02460818
45  -7.3464358  1.17296583
46  -7.3330908 -0.46917930
47  -8.6131366  0.83232288
50  -8.2215800 -0.14918470
51   1.8256208 -0.38448072
64   2.9026162 -0.56079106
66   1.4637874 -0.26431421
69   3.7577435 -1.86116228
70   1.1361477 -1.50155059
71   3.8827567  1.22906354
73   4.1692179 -1.46232092
77   2.8426912 -1.25559822
81   1.1230287 -1.61595346
83   0.9547656 -0.89366397
85   2.9951983  0.45345143
89   1.3563801  0.05717572
94   0.1395464 -1.52100735
95   2.0803720 -0.69753251
97   1.6919755 -0.29179654
99  -0.6967476 -0.67654495
100  1.6129364 -0.46383856
101  8.2517068  2.33297651
103  6.8760458  0.19728701
108  7.0477498 -1.16763040
111  4.7539370  1.22387159
119 10.0741201 -1.15217400
120  5.0909701 -2.04411170
125  6.1754568  1.24219739
127  4.3299493  0.12939584
128  4.3136703  0.59076457
129  6.9391309  0.31632256
131  6.9354478 -1.0

In [53]:
library(gmodels)

In [55]:
CrossTable(x=testLabels, y = testpred$class, prop.chisq=FALSE)


 
   Cell Contents
|-------------------------|
|                       N |
|           N / Row Total |
|           N / Col Total |
|         N / Table Total |
|-------------------------|

 
Total Observations in Table:  50 

 
             | testpred$class 
  testLabels |     setosa | versicolor |  virginica |  Row Total | 
-------------|------------|------------|------------|------------|
      setosa |         14 |          0 |          0 |         14 | 
             |      1.000 |      0.000 |      0.000 |      0.280 | 
             |      1.000 |      0.000 |      0.000 |            | 
             |      0.280 |      0.000 |      0.000 |            | 
-------------|------------|------------|------------|------------|
  versicolor |          0 |         16 |          1 |         17 | 
             |      0.000 |      0.941 |      0.059 |      0.340 | 
             |      0.000 |      1.000 |      0.050 |            | 
             |      0.000 |      0.320 |      0.020 |          

### 정확도 : 49/50 --> 98%
### versicolor 를 virginica로 잘못 예측
### 오분류율 : 1/50 --> 2%